### PR TITLE
Pre-Augmentations

### PR DESCRIPTION
**Pull request by kevinconka**
## What?

Added **pre-augmentation feature** to YOLOv5 that applies random cropping before the main data augmentation pipeline, creating a two-stage augmentation system.

### Extra tasks

- Move `train(..., im_compression_prob=0.8)` to the hyperparameters file as `compression=0.8`
- Delete unused scripts folder and `.sh` files
- Update `torch.cuda.amp` calls since there was an annoying deprecation warning.

## Why?

When training on high-res images, tiny objects often get _shrunk away_ during downscaling. Pre-cropping preserves them by isolating regions where they occupy more relative space.

This lets us:
- Retain fine-grained detail **without massive input sizes**
- Potentially boost detection of small objects **during deployment**
- Reduce overfitting by **adding spatial variation** early

## How?

**Implementation:**
- New `PreAlbumentations` class containing a composed transform
- New `RandomCropV2` class which dynamically adjusts the crop ROI by taking into account both image size and the crop size.
- Control random crop pre-augmentation via the hyperparameters file
  ```yaml
   # pre-augmentations (before main augmentations)
   pre_crop: 0.8  # random crop (probability)
   ```

> [!NOTE]
> To be honest, the `PreAlbumentations` seems like an overkill and even gives YAGNI vibes, but... the intention is to have a structure where we can add more pre-augmentations such as raindrops or fog rather easy.

**Flow:**
```
Original Image → Random Crop (PreAug) → Main Augmentation → Model Input
```

## Backout plan

Set `pre_crop=0.0` or do not include it in your `hyps.yaml`

## Testing

Using `notebooks/dataloader.ipynb`

###  `pre_crop=0.0`

image

### `pre_crop=0.5`

image

### `pre_crop=1.0`

image

### CODE REVIEW

File: utils/albumextensions.py [hunk 1]
```diff
@@ -1,16 +1,58 @@
...
 from albumentations.augmentations.geometric import functional as F
-from albumentations.core.pydantic import InterpolationType, ScaleIntType
+from albumentations.core.pydantic import InterpolationType
+from albumentations.core.types import ScaleIntType
+from albumentations.augmentations.crops import functional as fcrops
 from albumentations.core.transforms_interface import BaseTransformInitSchema, DualTransform
 from pydantic import ValidationInfo, field_validator
 
 
+class RandomCropV2(A.RandomCrop):
```
Comment (Zwicii) [inline utils/albumextensions.py]: Im not a fan of the naming.
V2?? yes we used it for ahoy and dan and so on - but why here? there will never be a V3.

I would name it SafeRandomCrop or something more descriptive

SafeRandomCrop clearly communicates its behavior — it's a safer variant that avoids errors when the crop is too large.
RandomCropV2 is to vague: “V2” doesn't explain what changed

File: data/hyps/hyp.sea-ai-IR.yaml [hunk 2]
```diff
@@ -32,6 +32,10 @@ fliplr: 0.5  # image flip left-right (probability)
 mosaic: 1.0  # image mosaic (probability)
 mixup: 0.0  # image mixup (probability)
 copy_paste: 0.0  # segment copy-paste (probability)
+compression: 0.8  # image compression (probability)
```
Comment (Zwicii) [inline data/hyps/hyp.sea-ai-IR.yaml]: before default value was 0.9 - why now 0.8?

File: data/hyps/hyp.sea-ai.yaml [hunk 3]
```diff
@@ -32,6 +32,10 @@ fliplr: 0.5  # image flip left-right (probability)
 mosaic: 1.0  # image mosaic (probability)
 mixup: 0.0  # image mixup (probability)
 copy_paste: 0.0  # segment copy-paste (probability)
+compression: 0.8  # image compression (probability)
```
Comment (Zwicii) [inline data/hyps/hyp.sea-ai.yaml]: before default value was 0.9 - why now 0.8?

File: utils/dataloaders.py [hunk 4]
```diff
@@ -694,15 +693,15 @@ def __init__(
...
+            fcn = self.cache_images_to_disk if cache_images == "disk" else self.load_image_and_labels
             with ThreadPool(NUM_THREADS) as pool:
                 results = pool.imap(lambda i: (i, fcn(i)), self.indices)
                 pbar = tqdm(results, total=len(self.indices), bar_format=TQDM_BAR_FORMAT, disable=LOCAL_RANK > 0)
                 for i, x in pbar:
                     if cache_images == "disk":
                         b += self.npy_files[i].stat().st_size
                     else:  # 'ram'
-                        self.ims[i], self.im_hw0[i], self.im_hw[i] = x  # im, hw_orig, hw_resized = load_image(self, i)
+                        self.ims[i], self.im_hw0[i], self.im_hw[i], _, _ = x  # im, hw_orig, hw_resized = load_image(self, i)
```
Comment (Zwicii) [inline utils/dataloaders.py]: Maybe we can update the comment so that it is up to date?
`self.ims[i], self.im_hw0[i], self.im_hw[i], _, _ = x  # im, hw_orig, hw_resized, labels, segments = load_image_and_labels(self, i)`

**Review by Zwicii — APPROVED** [review]
Just a few minor things — otherwise, looks great thanks :)

**Review by Zwicii — DISMISSED** [review]
Just a few minor things — otherwise, looks great thanks :)

**Review by Zwicii — APPROVED** [review]
